mono/metadata/object.c

```diff
@@ -380,6 +382,18 @@ mono_runtime_class_init_full (MonoVTable *vtable, MonoError *error)
 		mono_error_set_exception_instance (error, get_type_init_exception_for_vtable (vtable));
 		return FALSE;
 	}
+	if (!mono_threads_try_begin_abort_protected_block ()) {
+		mono_type_initialization_unlock ();
+		MonoException *exc = mono_thread_interruption_checkpoint ();
+		if (!exc) {
+			if (retry_count > 10)
+				g_error ("Retry count is too big (%d), this should not happen", retry_count);
+			goto retry_init;
+		}
```

Comment:
There are a few problems:

1) we must not execute interrupt while holding the type initialization. As that's a leaf lock.
2) interruption can potentially return NULL... for reasons. Some valid, some just me not being super familiar with the code.

My concern is that the unknown parts of (2) could introduce a retry loop and that's what I'm trying to catch.